File: Pod/Classes/RDPageContentData/RDRemoteImageContentData/RDRemoteImageContentData.m
//
//  RDRemoteImageContentData.m
//  RDImageViewerController
//
//  Created by Akira Matsuda on 9/18/16.
//  Copyright © 2016 Akira Matsuda. All rights reserved.
//

#import "RDRemoteImageContentData.h"
#import "RDImageScrollView.h"
#import "rd_M2DURLConnectionOperation.h"

@interface RDRemoteImageContentData ()

@property (nonatomic, strong) RDImageScrollView *imageView;
@property (nonatomic, strong) rd_M2DURLConnectionOperation *operation;
@property (nonatomic, strong) NSURLRequest *request;
@property (nonatomic, assign) NSInteger pageIndex;
@property (nonatomic, strong) void(^lazyConfigurationHandler)(UIImage *image);

@end

@implementation RDRemoteImageContentData

- (instancetype)initWithRequest:(NSURLRequest *)request pageIndex:(NSInteger)index
{
	self = [super init];
	if (self) {
		self.request = request;
		self.pageIndex = index;
	}
	
	return self;
}

- (void)stopPreloading
{
	[self.operation stop];
}

#pragma mark - RDPageContentDataDelegate

+ (UIView * _Nonnull)contentViewWithFrame:(CGRect)frame
{
	RDImageScrollView *view = [[RDImageScrollView alloc] initWithFrame:frame];
	
	return view;
}

- (void)preload
{
	if (self.image == nil) {
		self.imageView = nil;
		__weak typeof(self) bself = self;
		self.operation = [[rd_M2DURLConnectionOperation alloc] initWithRequest:self.request completeBlock:^(rd_M2DURLConnectionOperation *operation, NSURLResponse *response, NSData *data, NSError *error) {
			if (bself.requestCompletionHandler) {
				bself.requestCompletionHandler(response, data, error);
			}
			
			UIImage *image = nil;
			if (bself.imageDecodeHandler) {
				image = bself.imageDecodeHandler(data);
			}
			else {
				image = [[UIImage alloc] initWithData:data];
			}
			
			bself.image = image;
			if (bself.lazyConfigurationHandler) {
				bself.lazyConfigurationHandler(image);
			}
		}];
		self.operation.configuration = self.configuration;
		[self.operation sendRequest];
	}
}

- (void)reload
{
	self.image = nil;
	[self preload];
}

- (void)configureForView:(UIView * _Nonnull)view
{
	[super configureForView:view];
	if (self.image == nil) {
		RDImageScrollView *imageView = (RDImageScrollView *)view;
		self.imageView = imageView;
		__weak typeof(self) bself = self;
		__weak typeof(imageView) bimageView = imageView;
		self.lazyConfigurationHandler = ^(UIImage *image) {
			if (bimageView == bself.imageView) {
				dispatch_async(dispatch_get_main_queue(), ^{
					bimageView.image = image;
				});
			}
		};
	}
}

@end

File: Pod/Classes/RDRemoteImageContentData.swift
//
//  RDRemoteImageContentData.swift
//  Pods-RDImageViewerController
//
//  Created by Akira Matsuda on 2019/04/07.
//

import UIKit

class RDRemoteImageContentData: RDImageContentData {
    private var tempolaryImageView: RDImageScrollView?
    private let request: URLRequest
    private let session: URLSession
    private var task: URLSessionTask?
    private var lazyConfigurationHandler: ((UIImage?) -> Void)?
    var configuration: URLSessionConfiguration?
    var completionHandler: ((Data?, URLResponse?, Error?) -> Void)?
    var imageDecodeHandler: ((Data) -> UIImage?)?
    
    init(request: URLRequest, session: URLSession) {
        self.session = session
        self.request = request
        super.init(imageName: "")
    }
    
    override func stopPreload() {
        if let task = task {
            task.cancel()
        }
    }
    
    override func contentView(frame: CGRect) -> UIView {
        return RDImageScrollView(frame: frame)
    }
    
    override func preload() {
        if image == nil {
            task = session.dataTask(with: request, completionHandler: { [weak self] (data, response, error) in
                guard let weakSelf = self, let data = data else {
                    return
                }
                if let handler = weakSelf.completionHandler {
                    handler(data, response, error)
                }
                if let decodeHandler = weakSelf.imageDecodeHandler {
                    weakSelf.image = decodeHandler(data)
                }
                else {
                    weakSelf.image = UIImage(data: data)
                }
                
                if let handler = weakSelf.lazyConfigurationHandler {
                    handler(weakSelf.image)
                }
            })
            
            task?.resume()
        }
    }
    
    override func reload() {
        image = nil
        preload()
    }
    
    override func configure(view: UIView) {
        super.configure(view: view)
        if image == nil {
            let imageView = view as! RDImageScrollView
            tempolaryImageView = imageView
            lazyConfigurationHandler = { [weak self] (image) in
                guard let weakSelf = self else {
                    return
                }
                if weakSelf.tempolaryImageView == imageView {
                    DispatchQueue.main.async {
                        weakSelf.tempolaryImageView?.image = image
                    }
                }
            }
        }
    }
}
